File: src/rosbot_py_examples/rosbot_py_examples/simple_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SimplePublisher(Node):
    """ROS 2 node that publishes incrementing greeting messages on /chatter."""

    def __init__(self):
        # Initialize the node with the ROS graph name `simple_publisher`.
        super().__init__('simple_publisher')

        # Create a String publisher:
        # - message type: std_msgs/String
        # - topic name: chatter
        # - queue depth: 10
        self.publisher_ = self.create_publisher(String, 'chatter', 10)

        # Internal state used to number each message.
        self.counter = 0

        # Publish once per second (1 Hz) using a periodic timer callback.
        self.frequency = 1.0  # Publish at 1 Hz
        self.timer = self.create_timer(1 / self.frequency, self.timer_callback)

        # Log startup status for visibility in the console.
        self.get_logger().info('Chatter publisher started, publishing at %f Hz' % self.frequency)

    def timer_callback(self):
        # Build and publish the next message, then increment the counter.
        msg = String()
        msg.data = 'Hello, world! #%d' % self.counter
        self.publisher_.publish(msg)
        self.counter += 1


def main():
    # Standard ROS 2 lifecycle:
    # 1) initialize rclpy
    # 2) create and spin node (process callbacks)
    # 3) clean up node and shutdown rclpy on exit
    rclpy.init()
    simple_publisher = SimplePublisher()
    rclpy.spin(simple_publisher)
    simple_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/rosbot_cpp_examples/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <chrono>

using namespace std::chrono_literals;

// SimplePublisher publishes a "Hello, ROS2!" message with an incrementing count.
class SimplePublisher : public rclcpp::Node
{
public:
    // Construct the node and set up communication primitives.
    SimplePublisher(): Node("simple_publisher")
    {
        // Create a publisher on the "chatter" topic with queue depth 10.
        publisher_ = create_publisher<std_msgs::msg::String>("chatter", 10);
        // Trigger timer_callback once per second.
        timer_ = create_wall_timer(1s, std::bind(&SimplePublisher::timer_callback, this));

        // Log that the node is ready.
        RCLCPP_INFO(get_logger(), "Simple Publisher node has been started.");
    }

private:
    // Counter appended to each published message.
    unsigned int count_ = 0;
    // ROS 2 publisher for std_msgs/String messages.
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    // Periodic timer used to publish at a fixed rate.
    rclcpp::TimerBase::SharedPtr timer_;

    // Compose and publish one message on each timer event.
    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello, ROS2! Count: " + std::to_string(count_++);
        publisher_->publish(message);
    }
};

int main(int argc, char * argv[])
{
    // Initialize ROS 2, run the node, then shut down cleanly.
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimplePublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
